File: Javascript/129. Sum Root to Leaf Numbers.js
/*
You are given the root of a binary tree containing digits from 0 to 9 only.
Each root-to-leaf path in the tree represents a number.
For example, the root-to-leaf path 1 -> 2 -> 3 represents the number 123.
Return the total sum of all root-to-leaf numbers. Test cases are generated so that the answer will fit in a 32-bit integer.
A leaf node is a node with no children.
*/
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}
/**
 * 129. Sum Root to Leaf Numbers
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function(root) {
    /*
    // DFS Recursive
    const dfs = function(node, currentSum) {
        if (node === null) { return 0; }
        // I know the () aren't needed because PEDMAS, but I like them.
        // * 10 to shift the number one decimal to the left before adding.
        currentSum = ((currentSum * 10) + node.val);
        if (node.left === null && node.right === null) {
            return currentSum;
        }
        let leftSum = dfs(node.left, currentSum);
        let rightSum = dfs(node.right, currentSum);
        return leftSum + rightSum;
    }
    return dfs(root, 0);
    */
    
    // BFS Iterative
    if (root === null) { return 0; }
    
    let sum = 0;
    let queue = [[root, 0]];
    
    while (queue.length > 0) {
        let [node, currentSum] = queue.shift();
        // I know the () aren't needed because PEDMAS, but I like them.
        // * 10 to shift the number one decimal to the left before adding.        
        currentSum = (currentSum * 10) + node.val;
        
        if (!node.left && !node.right) {
            sum += currentSum;
        }
        
        if (node.left) { queue.push([node.left, currentSum]); }
        if (node.right) { queue.push([node.right, currentSum]); }
    }
    
    return sum;
    
};

/*
const root = new TreeNode(1,
    new TreeNode(2, null, null),
    new TreeNode(3, null, null)
); // Expect 25
*/
const root = new TreeNode(4,
    new TreeNode(9, 
        new TreeNode(5, null, null),
        new TreeNode(1, null, null)
    ),
    new TreeNode(0, null, null)
); // Expect 1026

console.log(sumNumbers(root));

File: Rust/129_Sum_Root_to_Leaf_Numbers.rs
/*
You are given the root of a binary tree containing digits from 0 to 9 only.
Each root-to-leaf path in the tree represents a number.
For example, the root-to-leaf path 1 -> 2 -> 3 represents the number 123.
Return the total sum of all root-to-leaf numbers. Test cases are generated so that the answer will fit in a 32-bit integer.
A leaf node is a node with no children.
*/
/* 129. Sum Root to Leaf Numbers */
use std::rc::Rc;
use std::cell::RefCell;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(
        val: i32, 
        left: Option<Rc<RefCell<TreeNode>>>, 
        right: Option<Rc<RefCell<TreeNode>>>
    ) -> Rc<RefCell<TreeNode>> {
        Rc::new(RefCell::new(TreeNode { val, left, right }))
    }
}
struct Solution;
impl Solution {
    pub fn sum_numbers(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        
        fn dfs(node: Option<Rc<RefCell<TreeNode>>>, current_sum: i32) -> i32 {
            if let Some(node) = node {
                let node = node.borrow();
                let new_sum = current_sum * 10 as i32 + node.val;
                
                if node.left.is_none() && node.right.is_none() {
                    return new_sum;
                }
                
                let left_sum = dfs(node.left.clone(), new_sum);
                let right_sum = dfs(node.right.clone(), new_sum);
                
                return left_sum + right_sum;
            }
            0
        }        
        
        dfs(root, 0)
    }
}

fn main() {
    let root = TreeNode::new(1, Some(TreeNode::new(2, None, None)), Some(TreeNode::new(3, None, None)));
    /*
    let root = TreeNode::new(
        4, 
        Some(TreeNode::new(9, 
            Some(TreeNode::new(5, None, None)),
            Some(TreeNode::new(1, None, None))
        )),
        Some(TreeNode::new(0, None, None)) 
    );
    */
    let result = Solution::sum_numbers(Some(root));
    println!("{:?}", result);
}